### PR TITLE
Feature: Visitors and registered users can add to products in cart

### PR DESCRIPTION
**Pull request by ssoonsl**
Hey guys!

This PR makes the following changes:
- Visitors can now add products to cart
  - A temporary shopping cart is created for the visitor when they attempt to add to their shopping carts. The shopping cart id is stored in the user's session hash.
  - To perform this functionality, `optional: true` had to be added to the `users` column in the Shopping Cart model. After merging to master, you likely have to drop the table and migrate again (but thanks to Ash's hard work at setting up Faker seeds, we can get all the data back easily)
- Visitors can now view their shopping carts
  - Because the controllers and views for visitors and users turned out to be more or less identical, I decided to combine their routes so both visitors and users use the `meow::` namespace
  - We can require user authentication at the orders controller so visitors are not allowed to make orders unless they sign up
- When a user is routed to their shopping cart, their session data is checked to make sure any items in the  session cart is added to the user's registered shopping cart
- All users can now update desired quantity values when viewing their shopping carts

Because we won't be meeting over the weekend, I'll add notes to the controllers to make it more readable. Apologies for the fat and ugly controllers! Much refactoring work still to be done.

Please review and let me know if you see any parts that can be improved! Thanks! :balloon: Hope you're having a great weekend!

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -3,11 +3,33 @@ class ApplicationController < ActionController::Base
...
+        visitor_cart = ShoppingCart.find(session['cart'])
+        line_items = visitor_cart.line_items
+        line_items.each do |line_item|
+          line_item.update(shopping_cart: current_user.shopping_cart)
+          line_item.save
+        end
+        visitor_cart.destroy
+        session.delete['cart']
+      end
+      current_user.shopping_cart
```
Comment (ssoonsl) [inline app/controllers/application_controller.rb:24]: This part checks if the user's session data has cart information in it. If it does, each line item is added to the user's registered shopping cart. The visitor cart is then destroyed and its session cart data cleared.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -3,11 +3,33 @@ class ApplicationController < ActionController::Base
...
+    else
+      unless session['cart']
+        cart = ShoppingCart.new
+        cart.save
+        session['cart'] = cart.id
+      end
+      ShoppingCart.find(session['cart'])
+    end
```
Comment (ssoonsl) [inline app/controllers/application_controller.rb:32]: If the user is not signed in, create a temporary shopping cart for them and store the shopping cart id in the session hash.

File: app/controllers/meow/line_items_controller.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  def update
+    @line_item = LineItem.find(params[:line_item_id])
+    difference = @line_item.quantity - line_item_params[:quantity].to_i
+    @line_item.quantity -= difference
+    @line_item.save
+    @product = @line_item.product
+    @product.quantity += difference
+    @product.save
+    redirect_to meow_shopping_cart_path
+  end
```
Comment (ssoonsl) [inline app/controllers/meow/line_items_controller.rb:37]: Update: This comes from the Shopping Cart view. Users are allowed to change their requested product quantities from the shopping cart.

The difference between the requested quantity and existing line item quantity is calculated and both the line item quantities and total product quantities are updated.

**Review by ashrafmansur — APPROVED** [review]
Hey Sarine! Thanks for doing up this part of the application. Code looks good to me although I'm not a 100% sure of how some parts of the application controller and the meow/line_items controller work but I think I'll seek clarification from you tomorrow. I'm good with merging this to master! :+1:

**Pull request merged**